soh/soh/Enhancements/crowd-control/CrowdControl.cpp

```diff
@@ -19,268 +21,301 @@ extern GlobalContext* gGlobalCtx;
+            if (socketsReady == 0) {
+                continue;
+            }
 
-                std::string jsonResponse = dataSend.dump();
-                SDLNet_TCP_Send(tcpsock, jsonResponse.c_str(), jsonResponse.size() + 1);
+            int len = SDLNet_TCP_Recv(tcpsock, &received, sizeof(received));
+            if (!len || !tcpsock || len == -1) {
+                SPDLOG_ERROR("[CrowdControl] SDLNet_TCP_Recv: {}", SDLNet_GetError());
+                break;
             }
-        } else if (effectResult == EffectResult::Retry && paused == false && packet->timeRemaining > 0) {
-            paused = true;
-            nlohmann::json dataSend;
-            dataSend["id"] = packet->packetId;
-            dataSend["type"] = 0;
-            dataSend["timeRemaining"] = packet->timeRemaining;
-            dataSend["status"] = EffectResult::Paused;
-
-            std::string jsonResponse = dataSend.dump();
-            SDLNet_TCP_Send(tcpsock, jsonResponse.c_str(), jsonResponse.size() + 1);
+
+            try {
+                Effect* incomingEffect = ParseMessage(received);
+
+                // If effect is a one off run, let's execute
+                if (!incomingEffect->timeRemaining) {
+                    EffectResult result =
+                        ExecuteEffect(incomingEffect->type.c_str(), incomingEffect->value, false);
+                    EmitMessage(tcpsock, incomingEffect->id, incomingEffect->timeRemaining, result);
+                } else {
+                    // check if a conflicting event is already active
+                    bool isConflictingEffectActive = false;
+                    for (Effect* pack : activeEffects) {
+                        if (pack != incomingEffect && pack->category == incomingEffect->category &&
+                            pack->id < incomingEffect->id) {
+                            isConflictingEffectActive = true;
+                            EmitMessage(tcpsock, incomingEffect->id, incomingEffect->timeRemaining,
+                                        EffectResult::Retry);
+
+                            break;
+                        }
+                    }
+
+                    // check if effect can be executed
+                    EffectResult result =
+                        ExecuteEffect(incomingEffect->type.c_str(), incomingEffect->value, true);
+                    if (result == EffectResult::Retry || result == EffectResult::Failure) {
+                        EmitMessage(tcpsock, incomingEffect->id, incomingEffect->timeRemaining, result);
+                        continue;
+                    }
+
+                    if (!isConflictingEffectActive) {
+                        activeEffectsMutex.lock();
+                        activeEffects.push_back(incomingEffect);
+                        activeEffectsMutex.unlock();
+                    }
+                }
+            } catch (nlohmann::json::parse_error& e) { SPDLOG_ERROR("Error parsing JSON: {}", e.what()); }
```

Comment:
Not a fan of this large `try ... catch` here.
Thoughts about making  `ParseMessage` return `nullptr` on error?
Then this could become:
```c++
Effect* incomingEffect = ParseMessage(received);
if (!incomingEffect) {
    continue;
}
```